packages/framework/client-logger/fluid-telemetry/src/test/containerTelemetryEndToEnd.spec.realsvc.ts

```diff
@@ -0,0 +1,215 @@
+import { SharedTree } from "@fluid-experimental/tree";
+import type Sinon from "sinon";
+import { spy } from "sinon";
+import { expect } from "chai";
+import { timeoutPromise } from "@fluidframework/test-utils/internal";
+import { startTelemetry } from "../factory/index.js";
+import {
+	ContainerTelemetryEventNames,
+	type ContainerConnectedTelemetry,
+	type ContainerDisconnectedTelemetry,
+	type ContainerDisposedTelemetry,
+	type IFluidTelemetry,
+	type ITelemetryConsumer,
+} from "../index.js";
+
+// This test suite creates an actual IFluidContainer and confirms events are fired with the expected names during the expected events
+
+describe("container telemetry E2E", () => {
+	let schema;
+	let tinyliciousClient: TinyliciousClient;
+	let telemetryConsumerConsumeSpy: Sinon.SinonSpy;
+	let testTelemetryConsumer: ITelemetryConsumer;
+
+	// Simple test class that will be used as the telemetry consumer.
+	class TestTelemetryConsumer implements ITelemetryConsumer {
+		public consume(event: IFluidTelemetry): void {
+			return;
+		}
+	}
+
+	before(() => {
+		tinyliciousClient = new TinyliciousClient({ connection: { port: 7070 } });
+		schema = {
+			initialObjects: {
+				sharedMap1: SharedTree,
+			},
+		};
+
+		testTelemetryConsumer = new TestTelemetryConsumer();
+		telemetryConsumerConsumeSpy = spy(testTelemetryConsumer, "consume");
+	});
+
+	beforeEach(() => {
+		// We need to reset the telemetry consumer spy for each test so the trackEvent calls
+		// from the last test don't bleed into the next one.
+		telemetryConsumerConsumeSpy.resetHistory();
+	});
+
+	it("IFluid container's 'connected' system event produces expected ContainerConnectedTelemetry using ITelemetryConsumer", async () => {
+		const { container } = await tinyliciousClient.createContainer(schema);
+
+		const containerId = await container.attach();
+		startTelemetry({
+			container,
+			containerId,
+			consumers: [testTelemetryConsumer],
+		});
+
+		// We don't know exactly when the given container events that we're looking for will fire so we have to
+		// wrap the container.on(...) event handler within a promise that will be awaited at the end of the test.
+		const { actualContainerTelemetry, expectedContainerTelemetry } = await timeoutPromise<{
+			actualContainerTelemetry: ContainerConnectedTelemetry;
+			expectedContainerTelemetry: ContainerConnectedTelemetry;
+		}>(
+			(resolve) => {
+				container.on("connected", () => {
+					// We are making an assumption here that the 'connected' event is the first and only event sent to the ITelemetryConsumer.
+					const containerTelemetryFromSpy = telemetryConsumerConsumeSpy.getCalls()[0]
+						.args[0] as ContainerConnectedTelemetry;
+
+					resolve({
+						actualContainerTelemetry: containerTelemetryFromSpy,
+						expectedContainerTelemetry: {
+							eventName: ContainerTelemetryEventNames.CONNECTED,
+							containerId,
+							containerInstanceId: containerTelemetryFromSpy.containerInstanceId,
```

Comment:
Can we not get the instance ID directly from our container variable? If we did, we could declare the entire "expected" object outside of the promise. But I may be missing something.

Reply:
I'm remembering that we don't expose the instance ID on `IFluidContainer`. That's a bummer. It doesn't seem quite right to use the "actual" value for it here in the "expected" object, but the alternative would be to drill into private members of the container, which likely isn't better.

Probably worth a brief comment explaining this. Alternatively, we could omit the instance IDs from the test comparisons. Either way seems reasonable to me.